Special case Rational * 1 and 1 * Rational

isuruf opened the pull request:
Avoids a call to igcd which takes a lot of time if the numerator and denominator of the Rational are large.

cc @rwst

isuruf committed:
Special case Rational * 1 and 1 * Rational

Avoids a call to igcd which takes a lot of time if the numerator and denominator of the Rational are large.

rwst commented:
+1

isuruf committed:
1 * Expr returns Expr only if Expr is a rational

1 * 0.0 returned 0 before and this commit keeps the old behaviour

isuruf commented:
Tests pass. Ready for review now

jksuom commented:
Looks good to me apart from a couple of remarks:

The tests for ``S.One`` could be in the beginning, before ``if isinstance(other, Rational):``.

``self is S.One`` will probably never occur in ``Rational.__mul__``. Instead, it could be in ``Integer.__mul__``.

isuruf commented:
I did try the tests before the `if isinstance(other, Rational):`, but I cannot return other if self is `S.One` because `1 * 0.0` should return `0`

`self is S.One` occurs in `Integer.__mul__` and if `other` is not a integer type then `Rational.__mul__` is called

jksuom commented:
> 1 * 0.0 should return 0

I think this is odd. Is there a compelling reason for change of type?

isuruf commented:
Actually no. When we implemented floats in SymEngine we made sure to return 0.0 which I think is the best approach. 

In SymPy however many tests depend on the fact that 1*0.0 = 0.

jksuom commented:
Thanks, I didn't know that. (I've seen that floats behave unexpectedly in some other connections.)

smichr commented:
> Is there a compelling reason for change of type?

I'm not sure that we can track the precision of Float zero. So there's no point in keeping a floating point 0 instead of a plain 0 (I think).

smichr commented:
>  I cannot return other if self is  S.One  because  1 * 0.0  should return  0 

you could return `other or S.Zero` which, if other is 0.0 will return the non-Float zero.

smichr commented on sympy/core/numbers.py:
```diff
@@ -1389,7 +1389,12 @@ def __sub__(self, other):
     @_sympifyit('other', NotImplemented)
     def __mul__(self, other):
         if isinstance(other, Rational):
```
```
if self is S.One:
    other, self = self, other
if other is S.One:
    return self or S.Zero
if isinstance(other, Rational):
    return Rational(self.p*other.p, self.q*other.q) 
...
```

isuruf committed:
Update Rational.__mul__

Fix is due to @smichr.
SymPy depends on the fact that 1 * 0.0 returns 0 instead of 0.0, therefore 1 * Expr returns Expr if non-zero and S.Zero if Expr is float zero.

smichr commented:
Sorry, I didn't know about that restriction on `self`. So how about

```python
    if self is S.One:
        return other or S.Zero
    if other is S.One:
        return self or S.Zero
    ...
```

isuruf committed:
Remove code assigning to self.

jksuom commented:
+1

smichr commented:
I have no idea why the ode test is failing. I've added a few changes and am testing it now.

smichr commented:
minor mods [here](https://github.com/isuruf/sympy/pull/1)

smichr commented:
I saved the log if anyone wants to see it. I then restarted the job.

isuruf commented:
Closed via #11019
